{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in /srv/conda/envs/notebook/lib/python3.7/site-packages (1.74)\r\n",
      "Requirement already satisfied: numpy in /srv/conda/envs/notebook/lib/python3.7/site-packages (from biopython) (1.17.1)\r\n"
     ]
    }
   ],
   "source": [
    "# install biopython on Jupyter server.\n",
    "import sys\n",
    "!python -m pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "from Bio import Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"dbsnp-user@nih.gov\" # provide your user email \n",
    "# RECOMMENDED: apply for API key from NCBI (https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/). \n",
    "# 10 queries per second with a valid API key, otherwise 3 queries per seconds are allowed for 'None'\n",
    "Entrez.api_key = None\n",
    "\n",
    "# dbSNP supported query terms (https://www.ncbi.nlm.nih.gov/snp/docs/entrez_help/) can be build and test online using web query builder (https://www.ncbi.nlm.nih.gov/snp/advanced) \n",
    "# esearch handle\n",
    "eShandle = Entrez.esearch(db=\"snp\",  # search dbSNP\n",
    "                          #complex query for missense and pathogenic variants in LPL gene with global MAF betweeen 0 and 0.01.\n",
    "                          term='LPL[All Fields] AND pathogenic[Clinical_Significance] AND missense variant[Function_Class] AND (00000.0000[GLOBAL_MAF] : 00000.0100[GLOBAL_MAF])', \n",
    "                          usehistory=\"y\", #cache result on server for download in batches\n",
    "                          retmax=20 # return 20 RSID max\n",
    "                         )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# get esearch result\n",
    "eSresult = Entrez.read(eShandle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Count : 5\n",
      "RetMax : 5\n",
      "RetStart : 0\n",
      "QueryKey : 1\n",
      "WebEnv : NCID_1_3075896_130.14.18.97_9001_1567108675_833541832_0MetA0_S_MegaStore\n",
      "IdList : ['386571803', '118204057', '52818902', '17850737', '268']\n",
      "TranslationSet : []\n",
      "TranslationStack : [{'Term': 'LPL[All Fields]', 'Field': 'All Fields', 'Count': '21533', 'Explode': 'N'}, {'Term': 'pathogenic[Clinical_Significance]', 'Field': 'Clinical_Significance', 'Count': '70373', 'Explode': 'N'}, 'AND', {'Term': 'missense variant[Function_Class]', 'Field': 'Function_Class', 'Count': '7741037', 'Explode': 'N'}, 'AND', {'Term': '00000.0000[GLOBAL_MAF]', 'Field': 'GLOBAL_MAF', 'Count': '0', 'Explode': 'N'}, {'Term': '00000.0100[GLOBAL_MAF]', 'Field': 'GLOBAL_MAF', 'Count': '0', 'Explode': 'N'}, 'RANGE', 'GROUP', 'AND']\n",
      "QueryTranslation : LPL[All Fields] AND pathogenic[Clinical_Significance] AND missense variant[Function_Class] AND (00000.0000[GLOBAL_MAF] : 00000.0100[GLOBAL_MAF])\n"
     ]
    }
   ],
   "source": [
    "# review results \n",
    "for k in eSresult:\n",
    "    print (k, \":\", eSresult[k])\n",
    "    \n",
    "#Output: Web environment (&WebEnv) and query key (&query_key) parameters specifying the location on the Entrez history server of the list of UIDs matching the Entrez query\n",
    "#https://www.ncbi.nlm.nih.gov/books/NBK25500/#chapter1.Storing_Search_Results\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get result RSIDs list 'Idlist'\n",
    "# total rs count \n",
    "rslist = (eSresult['IdList'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "386571803\n",
      "118204057\n",
      "52818902\n",
      "17850737\n",
      "268\n"
     ]
    }
   ],
   "source": [
    "# retmax = 20 so print only 20 RSIDs\n",
    "# additional results can be retrieved by batches\n",
    "# download in batches example http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc139 or see below.\n",
    "for rs in rslist:\n",
    "    print(rs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the WebEnv session cookie, and the QueryKey:\n",
    "\n",
    "webenv = eSresult[\"WebEnv\"]\n",
    "query_key = eSresult[\"QueryKey\"]\n",
    "total_count = int(eSresult[\"Count\"])\n",
    "query_key = eSresult[\"QueryKey\"]\n",
    "retmax = 2 # return 2 rs per batch example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Going to download record 1 to 2\n",
      "<?xml version=\"1.0\" ?>\n",
      "<ExchangeSet xmlns:xsi=\"https://www.w3.org/2001/XMLSchema-instance\" xmlns=\"https://www.ncbi.nlm.nih.gov/SNP/docsum\" xsi:schemaLocation=\"https://www.ncbi.nlm.nih.gov/SNP/docsum ftp://ftp.ncbi.nlm.nih.gov/snp/specs/docsum_eutils.xsd\" ><DocumentSummary uid=\"386571803\"><SNP_ID>268</SNP_ID><ALLELE_ORIGIN/><GLOBAL_MAFS><MAF><STUDY>1000Genomes</STUDY><FREQ>G=0.0052/26</FREQ></MAF><MAF><STUDY>ALSPAC</STUDY><FREQ>G=0.0187/72</FREQ></MAF><MAF><STUDY>Estonian</STUDY><FREQ>G=0.0239/107</FREQ></MAF><MAF><STUDY>ExAC</STUDY><FREQ>G=0.0134/1622</FREQ></MAF><MAF><STUDY>GnomAD</STUDY><FREQ>G=0.0145/455</FREQ></MAF><MAF><STUDY>GnomAD_exomes</STUDY><FREQ>G=0.0128/3212</FREQ></MAF><MAF><STUDY>NorthernSweden</STUDY><FREQ>G=0.0283/17</FREQ></MAF><MAF><STUDY>PAGE_STUDY</STUDY><FREQ>G=0.0050/391</FREQ></MAF><MAF><STUDY>TOPMED</STUDY><FREQ>G=0.0109/1370</FREQ></MAF><MAF><STUDY>TWINSUK</STUDY><FREQ>G=0.0200/74</FREQ></MAF></GLOBAL_MAFS><GLOBAL_POPULATION/><GLOBAL_SAMPLESIZE>0</GLOBAL_SAMPLESIZE><SUSPECTED/><CLINICAL_SIGNIFICANCE>pathogenic,risk-factor</CLINICAL_SIGNIFICANCE><GENES><GENE_E><NAME>LPL</NAME><GENE_ID>4023</GENE_ID></GENE_E></GENES><ACC>NC_000008.11</ACC><CHR>8</CHR><HANDLE>DEBNICK,PERLEGEN,MGC_GENOME_DIFF,APPLERA_GI,PERLEGEN,AFFY,ILLUMINA,BCMHGSC_JDW,ENSEMBL,ILLUMINA,PAGE_STUDY,BCM-HGSC-SUB,1000GENOMES,1000GENOMES,1000GENOMES,OMICIA,ILLUMINA,OMIM-CURATED-RECORDS,1000GENOMES,NHLBI-ESP,1000GENOMES,EXOME_CHIP,CLINSEQ_SNP,ILLUMINA,ILLUMINA,ILLUMINA,EVA-GONL,1000GENOMES,EVA_GENOME_DK,EVA_FINRISK,EVA_DECODE,EVA_UK10K_ALSPAC,EVA_UK10K_TWINSUK,EVA_EXAC,EVA_MGP,EVA_SVP,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,JJLAB,HUMAN_LONGEVITY,TOPMED,ILLUMINA,ILLUMINA,GNOMAD,GNOMAD,GNOMAD,AFFY,AFFY,SWEGEN,ILLUMINA,ILLUMINA,CSHL,TOPMED,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,EGCUT_WGS,EVA_DECODE,ILLUMINA,ACPOP,ILLUMINA,ILLUMINA,ILLUMINA,EVA,PAGE_CC,ILLUMINA</HANDLE><FXN_CLASS>coding_sequence_variant,missense_variant</FXN_CLASS><VALIDATED>by-frequency,by-cluster</VALIDATED><DOCSUM>HGVS=NC_000008.11:g.19956018A&gt;G,NC_000008.10:g.19813529A&gt;G,NG_008855.2:g.59302A&gt;G,NG_008855.1:g.21948A&gt;G,NM_000237.3:c.953A&gt;G,NM_000237.2:c.953A&gt;G,NP_000228.1:p.Asn318Ser|SEQ=[A/G]|GENE=LPL:4023</DOCSUM><TAX_ID>9606</TAX_ID><ORIG_BUILD>36</ORIG_BUILD><UPD_BUILD>153</UPD_BUILD><CREATEDATE>2000/09/19 17:02</CREATEDATE><UPDATEDATE>2019/07/13 14:17</UPDATEDATE><SS>268,23712753,28505039,48420135,69043148,74819581,74905524,93851529,143320833,173427532,181341869,206439351,217321419,217397623,217418282,244238715,244291490,252841599,334734988,342253784,490960919,491410891,491921986,537115398,780867933,783552867,985272617,1328915147,1582593752,1584057346,1594862271,1620133702,1663127735,1689111573,1711194704,1713021090,1752723237,1917826322,1946231540,1946231541,1959093877,1959093878,2024980549,2301287943,2470945845,2634720445,2711132106,2737022388,2748007727,2864092854,2985433050,2986076176,3002804432,3022826073,3022826074,3348082024,3555882704,3625947308,3630013637,3630013638,3635162175,3638748368,3640869465,3643680166,3644964714,3644964715,3653367027,3653367028,3654194864,3670484468,3721555439,3726520351,3735467033,3744302894,3744577751,3745461958,3767717753,3771428683,3772954551</SS><ALLELE>R</ALLELE><SNP_CLASS>snv</SNP_CLASS><CHRPOS>8:19956018</CHRPOS><TEXT>MergedRs=268</TEXT><SNP_ID_SORT>0000000268</SNP_ID_SORT><CLINICAL_SORT>1</CLINICAL_SORT><CITED_SORT/><CHRPOS_SORT>0019956018</CHRPOS_SORT><MERGED_SORT>1</MERGED_SORT></DocumentSummary>\n",
      "<DocumentSummary uid=\"118204057\"><SNP_ID>118204057</SNP_ID><ALLELE_ORIGIN/><GLOBAL_MAFS><MAF><STUDY>1000Genomes</STUDY><FREQ>A=0.0002/1</FREQ></MAF><MAF><STUDY>ALSPAC</STUDY><FREQ>A=0.0005/2</FREQ></MAF><MAF><STUDY>ExAC</STUDY><FREQ>A=0.0001/16</FREQ></MAF><MAF><STUDY>GnomAD</STUDY><FREQ>A=0.0001/2</FREQ></MAF><MAF><STUDY>GnomAD_exomes</STUDY><FREQ>A=0.0002/48</FREQ></MAF><MAF><STUDY>PAGE_STUDY</STUDY><FREQ>A=0.0002/13</FREQ></MAF><MAF><STUDY>TOPMED</STUDY><FREQ>A=0.0003/32</FREQ></MAF><MAF><STUDY>TWINSUK</STUDY><FREQ>A=0.0003/1</FREQ></MAF></GLOBAL_MAFS><GLOBAL_POPULATION/><GLOBAL_SAMPLESIZE>0</GLOBAL_SAMPLESIZE><SUSPECTED/><CLINICAL_SIGNIFICANCE>pathogenic</CLINICAL_SIGNIFICANCE><GENES><GENE_E><NAME>LPL</NAME><GENE_ID>4023</GENE_ID></GENE_E></GENES><ACC>NC_000008.11</ACC><CHR>8</CHR><HANDLE>OMIM-CURATED-RECORDS,NHLBI-ESP,EXOME_CHIP,ILLUMINA,ILLUMINA,1000GENOMES,EVA_UK10K_ALSPAC,EVA_UK10K_TWINSUK,EVA_EXAC,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,HUMAN_LONGEVITY,TOPMED,GNOMAD,GNOMAD,GNOMAD,AFFY,AFFY,SWEGEN,ILLUMINA,TOPMED,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,PAGE_CC,ILLUMINA</HANDLE><FXN_CLASS>coding_sequence_variant,missense_variant</FXN_CLASS><VALIDATED>by-frequency,by-cluster</VALIDATED><DOCSUM>HGVS=NC_000008.11:g.19954222G&gt;A,NC_000008.11:g.19954222G&gt;C,NC_000008.10:g.19811733G&gt;A,NC_000008.10:g.19811733G&gt;C,NG_008855.2:g.57506G&gt;A,NG_008855.2:g.57506G&gt;C,NG_008855.1:g.20152G&gt;A,NG_008855.1:g.20152G&gt;C,NM_000237.3:c.644G&gt;A,NM_000237.3:c.644G&gt;C,NM_000237.2:c.644G&gt;A,NM_000237.2:c.644G&gt;C,NP_000228.1:p.Gly215Glu,NP_000228.1:p.Gly215Ala|SEQ=[G/A/C]|GENE=LPL:4023</DOCSUM><TAX_ID>9606</TAX_ID><ORIG_BUILD>132</ORIG_BUILD><UPD_BUILD>153</UPD_BUILD><CREATEDATE>2010/08/10 18:39</CREATEDATE><UPDATEDATE>2019/07/13 14:17</UPDATEDATE><SS>252841577,342253778,491410889,780867932,783552866,1328915074,1620133671,1663127704,1689111495,1752723234,1917826320,1946231533,1959093860,2301287816,2470945734,2737022275,2748007694,2864092706,2985433044,2986076165,3002804416,3022826051,3555882370,3630013630,3635162172,3640869462,3644964707,3653367005,3654194858,3726520329,3744577749,3745461955,3771428668,3772954548</SS><ALLELE>V</ALLELE><SNP_CLASS>snv</SNP_CLASS><CHRPOS>8:19954222</CHRPOS><TEXT/><SNP_ID_SORT>0118204057</SNP_ID_SORT><CLINICAL_SORT>1</CLINICAL_SORT><CITED_SORT/><CHRPOS_SORT>0019954222</CHRPOS_SORT><MERGED_SORT>0</MERGED_SORT></DocumentSummary>\n",
      "</ExchangeSet>\n",
      "Going to download record 3 to 4\n",
      "<?xml version=\"1.0\" ?>\n",
      "<ExchangeSet xmlns:xsi=\"https://www.w3.org/2001/XMLSchema-instance\" xmlns=\"https://www.ncbi.nlm.nih.gov/SNP/docsum\" xsi:schemaLocation=\"https://www.ncbi.nlm.nih.gov/SNP/docsum ftp://ftp.ncbi.nlm.nih.gov/snp/specs/docsum_eutils.xsd\" ><DocumentSummary uid=\"52818902\"><SNP_ID>268</SNP_ID><ALLELE_ORIGIN/><GLOBAL_MAFS><MAF><STUDY>1000Genomes</STUDY><FREQ>G=0.0052/26</FREQ></MAF><MAF><STUDY>ALSPAC</STUDY><FREQ>G=0.0187/72</FREQ></MAF><MAF><STUDY>Estonian</STUDY><FREQ>G=0.0239/107</FREQ></MAF><MAF><STUDY>ExAC</STUDY><FREQ>G=0.0134/1622</FREQ></MAF><MAF><STUDY>GnomAD</STUDY><FREQ>G=0.0145/455</FREQ></MAF><MAF><STUDY>GnomAD_exomes</STUDY><FREQ>G=0.0128/3212</FREQ></MAF><MAF><STUDY>NorthernSweden</STUDY><FREQ>G=0.0283/17</FREQ></MAF><MAF><STUDY>PAGE_STUDY</STUDY><FREQ>G=0.0050/391</FREQ></MAF><MAF><STUDY>TOPMED</STUDY><FREQ>G=0.0109/1370</FREQ></MAF><MAF><STUDY>TWINSUK</STUDY><FREQ>G=0.0200/74</FREQ></MAF></GLOBAL_MAFS><GLOBAL_POPULATION/><GLOBAL_SAMPLESIZE>0</GLOBAL_SAMPLESIZE><SUSPECTED/><CLINICAL_SIGNIFICANCE>pathogenic,risk-factor</CLINICAL_SIGNIFICANCE><GENES><GENE_E><NAME>LPL</NAME><GENE_ID>4023</GENE_ID></GENE_E></GENES><ACC>NC_000008.11</ACC><CHR>8</CHR><HANDLE>DEBNICK,PERLEGEN,MGC_GENOME_DIFF,APPLERA_GI,PERLEGEN,AFFY,ILLUMINA,BCMHGSC_JDW,ENSEMBL,ILLUMINA,PAGE_STUDY,BCM-HGSC-SUB,1000GENOMES,1000GENOMES,1000GENOMES,OMICIA,ILLUMINA,OMIM-CURATED-RECORDS,1000GENOMES,NHLBI-ESP,1000GENOMES,EXOME_CHIP,CLINSEQ_SNP,ILLUMINA,ILLUMINA,ILLUMINA,EVA-GONL,1000GENOMES,EVA_GENOME_DK,EVA_FINRISK,EVA_DECODE,EVA_UK10K_ALSPAC,EVA_UK10K_TWINSUK,EVA_EXAC,EVA_MGP,EVA_SVP,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,JJLAB,HUMAN_LONGEVITY,TOPMED,ILLUMINA,ILLUMINA,GNOMAD,GNOMAD,GNOMAD,AFFY,AFFY,SWEGEN,ILLUMINA,ILLUMINA,CSHL,TOPMED,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,EGCUT_WGS,EVA_DECODE,ILLUMINA,ACPOP,ILLUMINA,ILLUMINA,ILLUMINA,EVA,PAGE_CC,ILLUMINA</HANDLE><FXN_CLASS>coding_sequence_variant,missense_variant</FXN_CLASS><VALIDATED>by-frequency,by-cluster</VALIDATED><DOCSUM>HGVS=NC_000008.11:g.19956018A&gt;G,NC_000008.10:g.19813529A&gt;G,NG_008855.2:g.59302A&gt;G,NG_008855.1:g.21948A&gt;G,NM_000237.3:c.953A&gt;G,NM_000237.2:c.953A&gt;G,NP_000228.1:p.Asn318Ser|SEQ=[A/G]|GENE=LPL:4023</DOCSUM><TAX_ID>9606</TAX_ID><ORIG_BUILD>36</ORIG_BUILD><UPD_BUILD>153</UPD_BUILD><CREATEDATE>2000/09/19 17:02</CREATEDATE><UPDATEDATE>2019/07/13 14:17</UPDATEDATE><SS>268,23712753,28505039,48420135,69043148,74819581,74905524,93851529,143320833,173427532,181341869,206439351,217321419,217397623,217418282,244238715,244291490,252841599,334734988,342253784,490960919,491410891,491921986,537115398,780867933,783552867,985272617,1328915147,1582593752,1584057346,1594862271,1620133702,1663127735,1689111573,1711194704,1713021090,1752723237,1917826322,1946231540,1946231541,1959093877,1959093878,2024980549,2301287943,2470945845,2634720445,2711132106,2737022388,2748007727,2864092854,2985433050,2986076176,3002804432,3022826073,3022826074,3348082024,3555882704,3625947308,3630013637,3630013638,3635162175,3638748368,3640869465,3643680166,3644964714,3644964715,3653367027,3653367028,3654194864,3670484468,3721555439,3726520351,3735467033,3744302894,3744577751,3745461958,3767717753,3771428683,3772954551</SS><ALLELE>R</ALLELE><SNP_CLASS>snv</SNP_CLASS><CHRPOS>8:19956018</CHRPOS><TEXT>MergedRs=268</TEXT><SNP_ID_SORT>0000000268</SNP_ID_SORT><CLINICAL_SORT>1</CLINICAL_SORT><CITED_SORT/><CHRPOS_SORT>0019956018</CHRPOS_SORT><MERGED_SORT>1</MERGED_SORT></DocumentSummary>\n",
      "<DocumentSummary uid=\"17850737\"><SNP_ID>268</SNP_ID><ALLELE_ORIGIN/><GLOBAL_MAFS><MAF><STUDY>1000Genomes</STUDY><FREQ>G=0.0052/26</FREQ></MAF><MAF><STUDY>ALSPAC</STUDY><FREQ>G=0.0187/72</FREQ></MAF><MAF><STUDY>Estonian</STUDY><FREQ>G=0.0239/107</FREQ></MAF><MAF><STUDY>ExAC</STUDY><FREQ>G=0.0134/1622</FREQ></MAF><MAF><STUDY>GnomAD</STUDY><FREQ>G=0.0145/455</FREQ></MAF><MAF><STUDY>GnomAD_exomes</STUDY><FREQ>G=0.0128/3212</FREQ></MAF><MAF><STUDY>NorthernSweden</STUDY><FREQ>G=0.0283/17</FREQ></MAF><MAF><STUDY>PAGE_STUDY</STUDY><FREQ>G=0.0050/391</FREQ></MAF><MAF><STUDY>TOPMED</STUDY><FREQ>G=0.0109/1370</FREQ></MAF><MAF><STUDY>TWINSUK</STUDY><FREQ>G=0.0200/74</FREQ></MAF></GLOBAL_MAFS><GLOBAL_POPULATION/><GLOBAL_SAMPLESIZE>0</GLOBAL_SAMPLESIZE><SUSPECTED/><CLINICAL_SIGNIFICANCE>pathogenic,risk-factor</CLINICAL_SIGNIFICANCE><GENES><GENE_E><NAME>LPL</NAME><GENE_ID>4023</GENE_ID></GENE_E></GENES><ACC>NC_000008.11</ACC><CHR>8</CHR><HANDLE>DEBNICK,PERLEGEN,MGC_GENOME_DIFF,APPLERA_GI,PERLEGEN,AFFY,ILLUMINA,BCMHGSC_JDW,ENSEMBL,ILLUMINA,PAGE_STUDY,BCM-HGSC-SUB,1000GENOMES,1000GENOMES,1000GENOMES,OMICIA,ILLUMINA,OMIM-CURATED-RECORDS,1000GENOMES,NHLBI-ESP,1000GENOMES,EXOME_CHIP,CLINSEQ_SNP,ILLUMINA,ILLUMINA,ILLUMINA,EVA-GONL,1000GENOMES,EVA_GENOME_DK,EVA_FINRISK,EVA_DECODE,EVA_UK10K_ALSPAC,EVA_UK10K_TWINSUK,EVA_EXAC,EVA_MGP,EVA_SVP,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,JJLAB,HUMAN_LONGEVITY,TOPMED,ILLUMINA,ILLUMINA,GNOMAD,GNOMAD,GNOMAD,AFFY,AFFY,SWEGEN,ILLUMINA,ILLUMINA,CSHL,TOPMED,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,EGCUT_WGS,EVA_DECODE,ILLUMINA,ACPOP,ILLUMINA,ILLUMINA,ILLUMINA,EVA,PAGE_CC,ILLUMINA</HANDLE><FXN_CLASS>coding_sequence_variant,missense_variant</FXN_CLASS><VALIDATED>by-frequency,by-cluster</VALIDATED><DOCSUM>HGVS=NC_000008.11:g.19956018A&gt;G,NC_000008.10:g.19813529A&gt;G,NG_008855.2:g.59302A&gt;G,NG_008855.1:g.21948A&gt;G,NM_000237.3:c.953A&gt;G,NM_000237.2:c.953A&gt;G,NP_000228.1:p.Asn318Ser|SEQ=[A/G]|GENE=LPL:4023</DOCSUM><TAX_ID>9606</TAX_ID><ORIG_BUILD>36</ORIG_BUILD><UPD_BUILD>153</UPD_BUILD><CREATEDATE>2000/09/19 17:02</CREATEDATE><UPDATEDATE>2019/07/13 14:17</UPDATEDATE><SS>268,23712753,28505039,48420135,69043148,74819581,74905524,93851529,143320833,173427532,181341869,206439351,217321419,217397623,217418282,244238715,244291490,252841599,334734988,342253784,490960919,491410891,491921986,537115398,780867933,783552867,985272617,1328915147,1582593752,1584057346,1594862271,1620133702,1663127735,1689111573,1711194704,1713021090,1752723237,1917826322,1946231540,1946231541,1959093877,1959093878,2024980549,2301287943,2470945845,2634720445,2711132106,2737022388,2748007727,2864092854,2985433050,2986076176,3002804432,3022826073,3022826074,3348082024,3555882704,3625947308,3630013637,3630013638,3635162175,3638748368,3640869465,3643680166,3644964714,3644964715,3653367027,3653367028,3654194864,3670484468,3721555439,3726520351,3735467033,3744302894,3744577751,3745461958,3767717753,3771428683,3772954551</SS><ALLELE>R</ALLELE><SNP_CLASS>snv</SNP_CLASS><CHRPOS>8:19956018</CHRPOS><TEXT>MergedRs=268</TEXT><SNP_ID_SORT>0000000268</SNP_ID_SORT><CLINICAL_SORT>1</CLINICAL_SORT><CITED_SORT/><CHRPOS_SORT>0019956018</CHRPOS_SORT><MERGED_SORT>1</MERGED_SORT></DocumentSummary>\n",
      "</ExchangeSet>\n",
      "Going to download record 5 to 5\n",
      "<?xml version=\"1.0\" ?>\n",
      "<ExchangeSet xmlns:xsi=\"https://www.w3.org/2001/XMLSchema-instance\" xmlns=\"https://www.ncbi.nlm.nih.gov/SNP/docsum\" xsi:schemaLocation=\"https://www.ncbi.nlm.nih.gov/SNP/docsum ftp://ftp.ncbi.nlm.nih.gov/snp/specs/docsum_eutils.xsd\" ><DocumentSummary uid=\"268\"><SNP_ID>268</SNP_ID><ALLELE_ORIGIN/><GLOBAL_MAFS><MAF><STUDY>1000Genomes</STUDY><FREQ>G=0.0052/26</FREQ></MAF><MAF><STUDY>ALSPAC</STUDY><FREQ>G=0.0187/72</FREQ></MAF><MAF><STUDY>Estonian</STUDY><FREQ>G=0.0239/107</FREQ></MAF><MAF><STUDY>ExAC</STUDY><FREQ>G=0.0134/1622</FREQ></MAF><MAF><STUDY>GnomAD</STUDY><FREQ>G=0.0145/455</FREQ></MAF><MAF><STUDY>GnomAD_exomes</STUDY><FREQ>G=0.0128/3212</FREQ></MAF><MAF><STUDY>NorthernSweden</STUDY><FREQ>G=0.0283/17</FREQ></MAF><MAF><STUDY>PAGE_STUDY</STUDY><FREQ>G=0.0050/391</FREQ></MAF><MAF><STUDY>TOPMED</STUDY><FREQ>G=0.0109/1370</FREQ></MAF><MAF><STUDY>TWINSUK</STUDY><FREQ>G=0.0200/74</FREQ></MAF></GLOBAL_MAFS><GLOBAL_POPULATION/><GLOBAL_SAMPLESIZE>0</GLOBAL_SAMPLESIZE><SUSPECTED/><CLINICAL_SIGNIFICANCE>pathogenic,risk-factor</CLINICAL_SIGNIFICANCE><GENES><GENE_E><NAME>LPL</NAME><GENE_ID>4023</GENE_ID></GENE_E></GENES><ACC>NC_000008.11</ACC><CHR>8</CHR><HANDLE>DEBNICK,PERLEGEN,MGC_GENOME_DIFF,APPLERA_GI,PERLEGEN,AFFY,ILLUMINA,BCMHGSC_JDW,ENSEMBL,ILLUMINA,PAGE_STUDY,BCM-HGSC-SUB,1000GENOMES,1000GENOMES,1000GENOMES,OMICIA,ILLUMINA,OMIM-CURATED-RECORDS,1000GENOMES,NHLBI-ESP,1000GENOMES,EXOME_CHIP,CLINSEQ_SNP,ILLUMINA,ILLUMINA,ILLUMINA,EVA-GONL,1000GENOMES,EVA_GENOME_DK,EVA_FINRISK,EVA_DECODE,EVA_UK10K_ALSPAC,EVA_UK10K_TWINSUK,EVA_EXAC,EVA_MGP,EVA_SVP,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,JJLAB,HUMAN_LONGEVITY,TOPMED,ILLUMINA,ILLUMINA,GNOMAD,GNOMAD,GNOMAD,AFFY,AFFY,SWEGEN,ILLUMINA,ILLUMINA,CSHL,TOPMED,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,ILLUMINA,EGCUT_WGS,EVA_DECODE,ILLUMINA,ACPOP,ILLUMINA,ILLUMINA,ILLUMINA,EVA,PAGE_CC,ILLUMINA</HANDLE><FXN_CLASS>coding_sequence_variant,missense_variant</FXN_CLASS><VALIDATED>by-frequency,by-cluster</VALIDATED><DOCSUM>HGVS=NC_000008.11:g.19956018A&gt;G,NC_000008.10:g.19813529A&gt;G,NG_008855.2:g.59302A&gt;G,NG_008855.1:g.21948A&gt;G,NM_000237.3:c.953A&gt;G,NM_000237.2:c.953A&gt;G,NP_000228.1:p.Asn318Ser|SEQ=[A/G]|GENE=LPL:4023</DOCSUM><TAX_ID>9606</TAX_ID><ORIG_BUILD>36</ORIG_BUILD><UPD_BUILD>153</UPD_BUILD><CREATEDATE>2000/09/19 17:02</CREATEDATE><UPDATEDATE>2019/07/13 14:17</UPDATEDATE><SS>268,23712753,28505039,48420135,69043148,74819581,74905524,93851529,143320833,173427532,181341869,206439351,217321419,217397623,217418282,244238715,244291490,252841599,334734988,342253784,490960919,491410891,491921986,537115398,780867933,783552867,985272617,1328915147,1582593752,1584057346,1594862271,1620133702,1663127735,1689111573,1711194704,1713021090,1752723237,1917826322,1946231540,1946231541,1959093877,1959093878,2024980549,2301287943,2470945845,2634720445,2711132106,2737022388,2748007727,2864092854,2985433050,2986076176,3002804432,3022826073,3022826074,3348082024,3555882704,3625947308,3630013637,3630013638,3635162175,3638748368,3640869465,3643680166,3644964714,3644964715,3653367027,3653367028,3654194864,3670484468,3721555439,3726520351,3735467033,3744302894,3744577751,3745461958,3767717753,3771428683,3772954551</SS><ALLELE>R</ALLELE><SNP_CLASS>snv</SNP_CLASS><CHRPOS>8:19956018</CHRPOS><TEXT/><SNP_ID_SORT>0000000268</SNP_ID_SORT><CLINICAL_SORT>1</CLINICAL_SORT><CITED_SORT/><CHRPOS_SORT>0019956018</CHRPOS_SORT><MERGED_SORT>0</MERGED_SORT></DocumentSummary>\n",
      "</ExchangeSet>\n"
     ]
    }
   ],
   "source": [
    "# sample codes adopted with modifications from http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc139.\n",
    "fetch_count = 0\n",
    "for start in range(0, total_count, retmax):\n",
    "    end = min(total_count, start+retmax)\n",
    "    print(\"Going to download record %i to %i\" % (start+1, end))\n",
    "    attempt = 0\n",
    "    #fetch_count += 1\n",
    "    while (attempt < 3):\n",
    "        attempt += 1\n",
    "        try:\n",
    "            fetch_handle = Entrez.efetch(db=\"snp\",\n",
    "                                         #rettype=\"uilist\", #available types [uilist | xml (use retmode=xml))\n",
    "                                         retmode=\"xml\",\n",
    "                                         retstart=start,\n",
    "                                         retmax=retmax,\n",
    "                                         webenv=webenv,\n",
    "                                         query_key=query_key )\n",
    "        except HTTPError as err:\n",
    "            if 500 <= err.code <= 599:\n",
    "                print(\"Received error from server %s\" % err)\n",
    "                print(\"Attempt %i of 3\" % attempt)\n",
    "                time.sleep(15)\n",
    "            else:\n",
    "                raise\n",
    "    if (fetch_handle):\n",
    "        #print(fetch_handle)            \n",
    "        data = fetch_handle.read()\n",
    "        print(data)\n",
    "        fetch_handle.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
